### PR TITLE
Update okhttp library

### PR DESCRIPTION
**Pull request by hannesa2**
____

QA:

- [x] (1) App crashes after typing credentials https://github.com/owncloud/android-library/pull/276#issuecomment-551091312

### CODE REVIEW

**Comment by jesmrec** [comment]
Smoke test required here

**Comment by jesmrec** [comment]
### (1)

1. Update the `owncloud-android-library`with the current PR
2. Open the app
3. Type URL and credentials (basic auth)

Current:

App crashes. This is the stacktrace:

 `2019-11-07 14:57:41.947 5896-6175/com.owncloud.android.debug E/AndroidRuntime: FATAL EXCEPTION: AsyncTask #3
    Process: com.owncloud.android.debug, PID: 5896
    java.lang.RuntimeException: An error occurred while executing doInBackground()
        at android.os.AsyncTask$3.done(AsyncTask.java:325)
        at java.util.concurrent.FutureTask.finishCompletion(FutureTask.java:354)
        at java.util.concurrent.FutureTask.setException(FutureTask.java:223)
        at java.util.concurrent.FutureTask.run(FutureTask.java:242)
        at android.os.AsyncTask$SerialExecutor$1.run(AsyncTask.java:243)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1133)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:607)
        at java.lang.Thread.run(Thread.java:761)
     Caused by: java.lang.NoSuchMethodError: No static method parse(Ljava/lang/String;)Lokhttp3/internal/http/StatusLine; in class Lokhttp3/internal/http/StatusLine; or its super classes (declaration of 'okhttp3.internal.http.StatusLine' appears in /data/app/com.owncloud.android.debug-1/base.apk)
        at at.bitfire.dav4android.PropStat$Companion.parse(PropStat.kt:51)
        at at.bitfire.dav4android.Response$Companion.parse(Response.kt:144)
        at at.bitfire.dav4android.DavResource$processMultiStatus$2.invoke(DavResource.kt:458)
        at at.bitfire.dav4android.DavResource.processMultiStatus(DavResource.kt:477)
        at at.bitfire.dav4android.DavResource.processMultiStatus(DavResource.kt:428)
        at at.bitfire.dav4android.DavResource.propfind(DavResource.kt:313)
        at com.owncloud.android.lib.common.http.methods.webdav.PropfindMethod.onExecute(PropfindMethod.java:62)
        at com.owncloud.android.lib.common.http.methods.webdav.DavMethod.execute(DavMethod.java:66)
        at com.owncloud.android.lib.common.OwnCloudClient.executeHttpMethod(OwnCloudClient.java:114)
        at com.owncloud.android.lib.resources.files.ExistenceCheckRemoteOperation.run(ExistenceCheckRemoteOperation.java:99)
        at com.owncloud.android.lib.common.operations.RemoteOperation.runOperation(RemoteOperation.java:251)
        at com.owncloud.android.lib.common.operations.RemoteOperation.execute(RemoteOperation.java:213)
        at com.owncloud.android.authentication.AuthenticatorAsyncTask.doInBackground(AuthenticatorAsyncTask.java:71)
        at com.owncloud.android.authentication.AuthenticatorAsyncTask.doInBackground(AuthenticatorAsyncTask.java:39)
        at android.os.AsyncTask$2.call(AsyncTask.java:305)
        at java.util.concurrent.FutureTask.run(FutureTask.java:237)
        at android.os.AsyncTask$SerialExecutor$1.run(AsyncTask.java:243) 
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1133) 
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:607) 
        at java.lang.Thread.run(Thread.java:761) 
`

Expected: Regular behaviour

Nexus 6P
Android 7.1.2
App commit: `a9073baf` (current master)
Server: 10.2.1

@hannesa2

**Comment by hannesa2** [comment]
Ok, I'm able to reproduce it
```
     Caused by: java.lang.NoSuchMethodError: No static method parse(Ljava/lang/String;)Lokhttp3/internal/http/StatusLine; in class Lokhttp3/internal/http/StatusLine; or its super classes (declaration of 'okhttp3.internal.http.StatusLine' appears in /data/app/com.owncloud.android-IzuDdJOCicRJ-50XZ-BeuA==/base.apk)
        at at.bitfire.dav4android.PropStat$Companion.parse(PropStat.kt:51)
        at at.bitfire.dav4android.Response$Companion.parse(Response.kt:144)
        at at.bitfire.dav4android.DavResource$processMultiStatus$2.invoke(DavResource.kt:458)
        at at.bitfire.dav4android.DavResource.processMultiStatus(DavResource.kt:477)
        at at.bitfire.dav4android.DavResource.processMultiStatus(DavResource.kt:428)
        at at.bitfire.dav4android.DavResource.propfind(DavResource.kt:313)
        at com.owncloud.android.lib.common.http.methods.webdav.PropfindMethod.onExecute(PropfindMethod.java:62)
```
and you are using
`api 'com.gitlab.ownclouders:dav4android:oc_support_1.0.1'` created by @davigonz 
which is a outdated fork from [https://gitlab.com/bitfireAT/dav4jvm](https://gitlab.com/bitfireAT/dav4jvm) which still used old `okhttp3` library.

So I really warmly recommend to bring your changes to the upstream master of Ricky Hirner, to get rid of this less maintained fork, otherwise the chance is very big to be locked with this for ever

So I run into this trap accidentally but I'm out now with update this dav-fork

**Comment by hannesa2** [comment]
Back to this PR:
The solution is quite easy, simply remove the dependency , because it's provided by 
`api 'com.gitlab.ownclouders:dav4android:oc_support_1.0.1'`

@jesmrec Now on my tests, there is no more crash

The maintenance of `com.gitlab.ownclouders:dav4android` is an other story

**Comment by jesmrec** [comment]
> So I really warmly recommend to bring your changes to the upstream master of Ricky Hirner, to get rid of this less maintained fork, otherwise the chance is very big to be locked with this for ever

the original dav library does not contain some stuff in which oC stuff is involved (am i wrong, @davigonz?). We agreed with dav4droid people (now called davx5) to do this in that way in the latest oC conference. Of course, it needs some maintenance. 

Crash does not happen anymore, i will perform the smoke testing.

Will ask new reviews to @davigonz and @abelgardep due to the latest changes.

**Comment by michaelstingl** [comment]
> We agreed with dav4droid people (now called davx5) to do this in that way in the latest oC conference. Of course, it needs some maintenance.

There must be a more sustainable way to add oC specific changes to upstream `bitfireAT/dav4jvm`?

**Comment by hannesa2** [comment]
> There must be a more sustainable way to add oC specific changes to upstream `bitfireAT/dav4jvm`?

1. Best solution is to move anyhow changes to upstream master. 

2. As plan B, a periodical rebase on master is very helpful. (I tried to rebase it to upstream master, but it was not funny. I gave up, my motivation was not big enough)

Doing 2. I have a inspiration for interested people: I do on other repos this [shell script](https://gist.github.com/hannesa2/0e7f93441e46a15669bcc4e284548c69) to **inform me**, when I should do a rebase. 
I have this in my `crontab -e`

`59 06  * * * ~/gitUpstreamCheck.sh ~/git/bitcoin-wallet `

**Comment by michaelstingl** [comment]
> * Best solution is to move anyhow changes to upstream master.

Some parts weren't accepted by the upstream project, because "they're too specific for the oC app". Not sure how the upstream part and the oC specific part can be combined. Make two libraries? Propose an extension architecture to the upstream project? (WebDAV itself can be extended too)

**Comment by davigonz** [comment]
> > So I really warmly recommend to bring your changes to the upstream master of Ricky Hirner, to get rid of this less maintained fork, otherwise the chance is very big to be locked with this for ever

> the original dav library does not contain some stuff in which oC stuff is involved (am i wrong, @davigonz?

You're right, davx5 guys didn't want some of our changes, so we had to separate the changes that the oC library needed and changes that are more generic and could be used for everyone using their library, we put those changes to upstream.

> As plan B, a periodical rebase on master is very helpful. (I tried to rebase it to upstream master, but it was not funny. I gave up, my motivation was not big enough)

They performed a deep rearchitecting of the project a bit after we started to use their library and it was not an easy rebase already. We will see when we can resume the work here.

**Comment by michaelstingl** [comment]
> We will see when we can resume the work here.

New issues for this task?

**Comment by davigonz** [comment]
> New issues for this task?

https://github.com/owncloud/android-library/issues/280

**Comment by davigonz** [comment]
Since here we tried to directly update okhttp library to 4.2.2 and it was finally changed to get the OkHttp dependency from dav4android, we are not upgrading the OkHttp version but downgrading it (see below), so I wouldn't address the update here but in https://github.com/owncloud/android-library/issues/280, after bringing the latest dav4jvm changes to our fork.

Before this PR:
Screenshot_2019-11-13_at_12_45_37

After this PR:
Screenshot_2019-11-13_at_12_37_49_1

Thanks anyway for the contribution @hannesa2

File: owncloudComLibrary/build.gradle [hunk 1]
```diff
@@ -5,7 +5,6 @@ apply plugin: 'kotlin-kapt'
 apply plugin: 'kotlin-allopen'
 
 dependencies {
-    api 'com.squareup.okhttp3:okhttp:3.12.0'
```
Comment (hannesa2) [inline owncloudComLibrary/build.gradle]: @davigonz But anyway, I would remove this misleading and pointless dependency and merge this PR

**Comment by davigonz** [comment]
> But anyway, I would remove this misleading and pointless dependency and merge this PR

It's not pointless since we are using it in the library, we prepared and tested the library code to work with OkHttp 3.12. There's some cases in which the same dependency got from a library is used with a different version in a project and nothing happens, you might want to have different versions.

In our case, I see your point and I know that we can get it from dav4android library but as I said in my previous comment, we would be downgrading the okhttp version from 3.12 to 3.11, what is even worse than keeping that dependency in build.gradle, getting rid of it (with proper QA and so on) within https://github.com/owncloud/android-library/issues/280 issue and really upgrading the version.

**Comment by hannesa2** [comment]
I see your point, but then please update in `dav4android` to `3.12` and keep this repo clean.
This is the place where it should be placed, updating here makes the topic more complicated then it should be

**Pull request closed without merging**